internal/helm/chart.go

```diff
@@ -18,35 +18,38 @@ package helm
 
 import (
 	"fmt"
-	"io"
-	"os"
-	"path"
 
+	helmchart "helm.sh/helm/v3/pkg/chart"
 	"helm.sh/helm/v3/pkg/chartutil"
 )
 
-// OverwriteChartDefaultValues overwrites the chart default values file in the
-// given chartPath with the contents of the given valuesFile.
-func OverwriteChartDefaultValues(chartPath, valuesFile string) error {
+// OverwriteChartDefaultValues overwrites the chart default values file with the
+// contents of the given valuesFile.
+func OverwriteChartDefaultValues(chart *helmchart.Chart, valuesFile string) error {
 	if valuesFile == "" || valuesFile == chartutil.ValuesfileName {
 		return nil
 	}
-	srcPath := path.Join(chartPath, valuesFile)
-	if f, err := os.Stat(srcPath); os.IsNotExist(err) || !f.Mode().IsRegular() {
-		return fmt.Errorf("invalid values file path: %s", valuesFile)
-	}
-	src, err := os.Open(srcPath)
-	if err != nil {
-		return fmt.Errorf("failed to open values file '%s': %w", valuesFile, err)
+
+	// Find override file and retrieve contents
+	var valuesData []byte
+	for _, f := range chart.Files {
+		if f.Name == valuesFile {
+			valuesData = f.Data
+			break
+		}
 	}
-	defer src.Close()
-	t, err := os.OpenFile(path.Join(chartPath, chartutil.ValuesfileName), os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
-	if err != nil {
-		return fmt.Errorf("failed to open values file '%s': %w", chartutil.ValuesfileName, err)
+	if valuesData == nil {
+		return fmt.Errorf("failed to locate override values file: %s", valuesFile)
 	}
-	defer t.Close()
-	if _, err := io.Copy(t, src); err != nil {
-		return fmt.Errorf("failed to overwrite default values with '%s': %w", valuesFile, err)
+
+	// Replace current values file
+	for _, f := range chart.Files {
```

Comment:
I looked at taking an approach like this (but decided not to as it would result in a very minor improvement without the additions in this PR), and iirc there are multiple places in the object where a copy of the "original values" is stored.

We should overwrite all those occurrences, as we do not know what will happen to the object after the overwrite which may give unexpected results in the future.

Reply:
Thanks for the hint, I can definitely see why the implementation was dealing directly with files and indeed it does seem this approach would be better fitted for this new scenario.

It seems that the original `values.yaml` is taken from [`Raw`](https://github.com/helm/helm/blob/b458bdf5c1baded057a6fc9bc08d554974614607/pkg/chart/chart.go#L32-L36) [when saving the chart](https://github.com/helm/helm/blob/b458bdf5c1baded057a6fc9bc08d554974614607/pkg/chartutil/save.go#L186), however that is being [overridden](https://github.com/helm/helm/blob/b458bdf5c1baded057a6fc9bc08d554974614607/pkg/chartutil/save.go#L213) by what's in [`Files`](https://github.com/helm/helm/blob/b458bdf5c1baded057a6fc9bc08d554974614607/pkg/chart/chart.go#L43-L44). To be on the safe side we can most certainly override both, but what seems to matter the most is what's in `Files`.

Another place where values are stored (unmarshaled yaml as `chartutil.Values`) is the `Values` [field](https://github.com/helm/helm/blob/b458bdf5c1baded057a6fc9bc08d554974614607/pkg/chart/chart.go#L43-L44).

Will follow-up with a new set of changes.